crates/sparrow-batch/src/batch.rs

```diff
@@ -231,47 +261,58 @@ impl Batch {
     /// Returns the rows less than the given time and
     /// leaves the remaining rows in `self`.
     pub fn split_up_to(&mut self, time_exclusive: RowTime) -> Option<Batch> {
-        let Some(data) = &mut self.data else {
-            return None;
-        };
-
-        if time_exclusive < data.min_present_time {
-            // time_exclusive < min_present <= max_present.
-            // none of the rows in the batch should be taken.
-            return None;
-        }
+        if let Some(data) = &mut self.data {
+            if time_exclusive < data.min_present_time {
+                // time_exclusive < min_present <= max_present.
+                // none of the rows in the batch should be taken.
+                return None;
+            }
 
-        if data.max_present_time < time_exclusive {
-            // min_present <= max_present < time_exclusive
-            // all rows should be taken
-            return Some(Batch {
-                data: self.data.take(),
-                // Even though we took all the rows, the batch
-                // we return is only as complete as the original
-                // batch. There may be other batches after this
-                // that have equal rows less than time exclusive.
-                up_to_time: self.up_to_time,
-            });
-        }
+            if data.max_present_time < time_exclusive {
+                // min_present <= max_present < time_exclusive
+                // all rows should be taken
+                return Some(Batch {
+                    data: self.data.take(),
+                    // Even though we took all rows, we need to pick the
+                    // minimum up_to_time. If
+                    // 1) self.up_to_time < time_exclusive: the input may have more rows
+                    // between self.up_to_time and time_exclusive.
+                    // 2) self.up_to_time > time_exclusive: We requested a split at
+                    // time_exclusive, so we can't return a batch with a greater time.
+                    up_to_time: self.up_to_time.min(time_exclusive),
```

Comment:
This could go either way, I think. Technically, this definition is true. If we ask for a split at `time_exclusive` then we shouldn't produce an `up_to_time` greater than that. 

But, we could also assert that `time_exclusive` is always `<=` the `self.up_to_time`. The `Gatherer` should never ask for a split time greater than the `up_to_time` of the minimum batch.

Reply:
I'm not sure I understand if there is a question here, but:

1. It seems like if we can assert / narrow the cases this supports, that may be good for ensuring it is being used as we intended.
2.

Reply:
It was between using the minimum of `self.up_to_time` and the given `time_exclusive` or just using `self.up_to_time` and asserting that it's not being split at a time greater than `self.up_to_time`. 

This method could be used outside of the gatherer, which is why I chose to use the minimum, as it's more in line with the actual definition of `split_up_to`.

Reply:
I'm fine with either. If we rely on the property, it makes sense to add an assertion. I could also see either being fine. Perhaps a way to think about is -- how would this work if the watermark was outside of the batch. Specifically, once we do #836, how would our preferences change?

Reply:
I would think it's basically the same implementation -- if we recognize we are going to take the entire `batch`, we have to decide whether to take the minimum of the `Watermarked` batch's `up_to_time`, or the time we split at. 

The benefits of using the `WatermarkedBatch` don't help much here, as the structure is the same in places that need to handle the watermark. (e.g. `Optional<BatchInfo>`. vs. `Optional<Batch>`, and a watermark).

Reply:
Maybe. On the other hand:

1. The handling of `up_to_time` would land in the `MergePipeline` and/or `Gatherer`
2. The `Batch::split_up_to` would be exclusively focusing on splitting the data in the batch. It wouldn't have any logic on how to compute the `up_to_time`.

So in that world:

1. The `gatherer` would determine the `up_to_time` / `watermark` of the output.
2. The `Batch::split_up_to` would split the data at the request of the gatherer.

I *think* that would lead to: `Batch::split_up_to` should always split the batch at the requested time, and the gatherer would pick a properly watermark / time.

I guess it doesn't speak to what to do for the `split_up_to` for now.

Reply:
Hm, fair. I was imagining that the `WatermarkedBatch` would have a `split_up_to` method that was functionally the same as the current `Batch`'s, since it's likely we'll have mutiple places we'll need to split data + watermark (merge, shifts)